pkg/kubelet/kubelet.go

```diff
@@ -1982,6 +1985,7 @@ func (kl *Kubelet) HandlePodAdditions(pods []*api.Pod) {
 }
 
 func (kl *Kubelet) HandlePodUpdates(pods []*api.Pod) {
+	defer func(sourcesReadyValue bool) { kl.sourcesReadyForUse = sourcesReadyValue }(kl.sourcesReady())
```

Comment:
>I don't think this works correctly, as kl.sourcesReady() is evaluated a the binding to the deferred func, not at the end of the function.

yes, I want it evaluated to what `kl.sourcesReady()`  is, at the beginning of the function, not what value it is at the end of the function

>Also, I think we want to continue to dynamically evaluate that function. What is the value of statically storing it?

* The housekeeping routine, kills pods that are not in the podManager cache so long as `kl.sourcesReady()` is true. 
* `kl.sourcesReady()` is set to true asynchronously
* pods are added to the podManager by reading from the update/add channel.

The race is that if `kl.sourcesReady()` changes to true, and kubelet enters the housekeeping routine, it kills all the running pods. The housekeeping routine races with the reading from the update/add channel and processing it
Or to put it in other words, kubelet uses `kl.sourcesReady()` as all pods from all sources have been processed.